### PR TITLE
Declarative kernel module loading support for BSPs

### PR DESCRIPTION
**Pull request by ricmm**
This adds support to a declarative interface in the BSP JSON
for kernel modules that are provided by the BSP and that expose
the said modules to the application containers depending on the
presence of a named dtb or dtbo object as passed by the bootloader.

### CODE REVIEW

File: ctrl.c [hunk 1]
```diff
@@ -838,6 +841,36 @@ static struct pv_cmd* pv_ctrl_process_endpoint_and_reply(int req_fd,
...
+	} else if (pv_str_startswith(ENDPOINT_DRIVERS, strlen(ENDPOINT_DRIVERS), path)) {
+		driverkey = pv_ctrl_get_file_name(path, sizeof(ENDPOINT_DRIVERS), path_len);
+
+		if (!driverkey) {
+			pv_log(WARN, "HTTP request has bad driver alias");
+			pv_ctrl_write_error_response(req_fd, HTTP_STATUS_BAD_REQ, "Request has bad driver key name");
+			goto out;
+		}
+
+		if (!strncmp("PUT", method, method_len)) {
```
Comment (anibalportero) [inline ctrl.c:853]: I don't know if I like using PUT for this one.. this kind of unload/upload ops seems more suited to a POST similar to the one we use for commands, and then put the op and driver in a json in the body like "{"op":"LOCAL_RUN","payload":"4"}". Not very strong on this one, but I'd do it like that to keep consistency.

File: state.c [hunk 2]
```diff
@@ -99,6 +99,7 @@ struct pv_state* pv_state_new(const char *rev, state_spec_t spec)
 		dl_list_init(&s->jsons);
 		pv_state_init_groups(s);
 		dl_list_init(&s->conditions);
+		dl_list_init(&s->bsp.drivers);
```
Comment (anibalportero) [inline state.c:102]: is it necessary to clean this list in pv_state_free? A change in bsp/drivers.json will always trigger a reboot update, but, maybe for completion's sake

File: platforms.c [hunk 3]
```diff
@@ -236,6 +237,19 @@ void pv_platform_add_condition(struct pv_platform *p, struct pv_condition *c)
 	}
 }
 
+void pv_platform_add_driver(struct pv_platform *p, plat_driver_t type, char *value)
+{
+	struct pv_platform_driver *d = calloc(1, sizeof(struct pv_platform_driver));
```
Comment (anibalportero) [inline platforms.c:242]: where are we freeing up this?

File: drivers.h [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+	struct dl_list list;
+};
+
+const char *pv_drivers_state_str(char *match);
+int pv_drivers_state(char *match);
+int pv_drivers_load(char *match);
+int pv_drivers_unload(char *match);
+
+struct pv_driver* pv_drivers_add(struct pv_state *s, char *alias, int len, char **modules);
+void pv_drivers_remove(struct pv_driver *d);
```
Comment (anibalportero) [inline drivers.h:54]: maybe call this one from pv_state_free

File: platforms.h [hunk 5]
```diff
@@ -54,6 +68,7 @@ struct pv_platform {
 	struct pv_state *state;
 	bool mgmt;
 	bool updated;
+	struct dl_list drivers;
```
Comment (anibalportero) [inline platforms.h:71]: if we hare allocating and adding drivers to this list that is separated from the one in drivers.json. Are they intended to be separate lists? Would that mean that one cannot consult platform drivers from ctrl? Maybe also that drivers are going to be loaded in state but not in platform? Maybe I'm missing something

File: platforms.c (hunk 3 above)
Comment (asac) [reply platforms.c:252]: should be fixed now. please check @anibalportero

File: drivers.h (hunk 4 above)
Comment (asac) [reply drivers.h:56]: i think this is done now too. check it.

File: ctrl.c (hunk 1 above)
Comment (asac) [reply ctrl.c:864]: lets keep it at PUT for now.... there is not clear winner, in REST world a PUT would about an endpoint taht updates an existin resource, while a post would be about an endpoint where you can create new documents/resources. I think module loading is like activation and deactivation that REST APIs usually do with a PUT unless its about a activate/deactivate job resource that gets POSTED and then processed.

File: ctrl.c [hunk 6]
```diff
@@ -838,6 +844,67 @@ static struct pv_cmd* pv_ctrl_process_endpoint_and_reply(int req_fd,
...
+			if (!mgmt)
```
Comment (anibalportero) [inline ctrl.c:865]: so in the end, only mgmt can load and unload their own drivers, right?

File: drivers.h (hunk 4 above)
Comment (anibalportero) [reply drivers.h:57]: Looks good now. I guess we could also remove pv_drivers_remove as that's not used anywhere

**Pull request merged**